### PR TITLE
MAINT: Migrate from versioneer to setuptools_scm

### PR DESCRIPTION
**Pull request by oesteban**
After the experience with setuptools_scm, it seems a lightweight replacement for versioneer while it is really much easier to overcome the problem of missing the git history when building containers.

I propose this PR as a pilot before adopting setuptools_scm more widely.

### CODE REVIEW

**Comment by arokem** [comment]
+1 from me for scm_setuptools. We've started using that in pyAFQ and it does seem like a rather good way to go. The only quirk we've found is that you need to hack around it's versioning scheme if you want to upload to testpypi as part of the CI. But something like this does it: https://github.com/yeatmanlab/pyAFQ/blob/master/setup.py#L29-L43

**Comment by oesteban** [comment]
@arokem I see the problem:

```
HTTPError: 400 Client Error: '1.2.1.dev5+g1e9790be' is an invalid value for Version. Error: Can't use PEP 440 local versions. See https://packaging.python.org/specifications/core-metadata for url: https://test.pypi.org/legacy/
```
I guess we can keep going without the TestPypi attempt.

**Comment by arokem** [comment]
Yes. That's exactly the problem. You can back off uploads to TestPyPI, or use something like our hack to provide an acceptable version number.

File: pyproject.toml [hunk 1]
```diff
@@ -1,2 +1,10 @@
 [build-system]
-requires = ["setuptools >= 40.8.0", "wheel"]
+requires = ["setuptools ~= 42.0", "wheel", "setuptools_scm[toml] >= 3.4"]
```
Comment (effigies) [inline pyproject.toml]: Elsewhere it looks like you're permitting setuptools 44. I'm guessing this is a typo.

```suggestion
requires = ["setuptools >= 42.0", "wheel", "setuptools_scm[toml] >= 3.4"]
```

File: MANIFEST.in [hunk 2]
```diff
@@ -3,6 +3,8 @@ include LICENSE
...
 recursive-include sdcflows/data *
 
-# versioneer
-include versioneer.py
-include sdcflows/_version.py
+recursive-exclude .circleci/ *
+recursive-exclude .docker/ *
+recursive-exclude .github/ *
+
+exclude .codecov.yml .coveragerc .dockerignore .gitattributes .gitignore .pep8speaks.yml .travis.yml .zenodo.json Dockerfile
```
Comment (effigies) [inline MANIFEST.in]: Have these files been getting included? In general `setup.py` will only include Python files.

**Review by effigies — COMMENTED** [review]
I have no objection to this move and if it lets us stop using a custom variant of such machinery, then cool. I have also considered possibly having a nipy fork of versioneer (https://github.com/nipy/nibabel/issues/816).

I might check how it handles `git archive`, which will affect whether something like the following works:

```
pip install https://github.com/nipreps/sdcflows/archive/1.1.0.tar.gz
```

Versioneer handles archives of tags well, but normal dev versions poorly. Though no VCS-based versioning handles dev versions well for archives, since git doesn't have a way to interpolate the most recent tag with its template language.

File: MANIFEST.in (hunk 2 above)
Comment (oesteban) [reply MANIFEST.in]: Yes, if I comment that line out:
```Shell
$ python setup.py sdist
running sdist
running egg_info
writing sdcflows.egg-info/PKG-INFO
writing dependency_links to sdcflows.egg-info/dependency_links.txt
writing entry points to sdcflows.egg-info/entry_points.txt
writing requirements to sdcflows.egg-info/requires.txt
writing top-level names to sdcflows.egg-info/top_level.txt
reading manifest template 'MANIFEST.in'
writing manifest file 'sdcflows.egg-info/SOURCES.txt'
running check
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/_static
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/reference
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/sphinxext
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/tools
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows.egg-info
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/cli
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/interfaces
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/tests
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/viz
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
creating sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows/tests
copying files to sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401...
copying .codecov.yml -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying .coveragerc -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying .dockerignore -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying .gitattributes -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying .gitignore -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying .pep8speaks.yml -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying .zenodo.json -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying CHANGES.rst -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying Dockerfile -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying LICENSE -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying MANIFEST.in -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying README.rst -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying pyproject.toml -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying setup.cfg -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying setup.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401
copying docs/Makefile -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
copying docs/api.rst -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
copying docs/changes.rst -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
copying docs/conf.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
copying docs/index.rst -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
copying docs/installation.rst -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
copying docs/links.rst -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
copying docs/requirements.txt -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs
copying docs/_static/RTD-advanced-conf.png -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/_static
copying docs/_static/git.png -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/_static
copying docs/reference/.gitignore -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/reference
copying docs/sphinxext/docscrape.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/sphinxext
copying docs/sphinxext/docscrape_sphinx.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/sphinxext
copying docs/sphinxext/github.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/sphinxext
copying docs/sphinxext/math_dollar.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/sphinxext
copying docs/sphinxext/numpydoc.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/sphinxext
copying docs/tools/LICENSE.txt -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/tools
copying docs/tools/apigen.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/tools
copying docs/tools/buildmodref.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/docs/tools
copying sdcflows/__about__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows
copying sdcflows/__init__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows
copying sdcflows/_version.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows
copying sdcflows/conftest.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows
copying sdcflows.egg-info/PKG-INFO -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows.egg-info
copying sdcflows.egg-info/SOURCES.txt -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows.egg-info
copying sdcflows.egg-info/dependency_links.txt -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows.egg-info
copying sdcflows.egg-info/entry_points.txt -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows.egg-info
copying sdcflows.egg-info/requires.txt -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows.egg-info
copying sdcflows.egg-info/top_level.txt -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows.egg-info
copying sdcflows/cli/__init__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/cli
copying sdcflows/cli/run.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/cli
copying sdcflows/data/__init__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
copying sdcflows/data/affine.json -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
copying sdcflows/data/fmap-any_registration.json -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
copying sdcflows/data/fmap-any_registration_testing.json -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
copying sdcflows/data/fmap_atlas.nii.gz -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
copying sdcflows/data/fmap_atlas_2_MNI152NLin2009cAsym_affine.mat -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
copying sdcflows/data/susceptibility_syn.json -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
copying sdcflows/data/translation_rigid.json -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/data
copying sdcflows/interfaces/__init__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/interfaces
copying sdcflows/interfaces/fmap.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/interfaces
copying sdcflows/interfaces/reportlets.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/interfaces
copying sdcflows/tests/__init__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/tests
copying sdcflows/tests/test_version.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/tests
copying sdcflows/viz/__init__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/viz
copying sdcflows/viz/utils.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/viz
copying sdcflows/workflows/__init__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/base.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/fmap.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/gre.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/outputs.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/pepolar.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/phdiff.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/syn.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/unwarp.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows
copying sdcflows/workflows/tests/__init__.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows/tests
copying sdcflows/workflows/tests/test_base.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows/tests
copying sdcflows/workflows/tests/test_pepolar.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows/tests
copying sdcflows/workflows/tests/test_phdiff.py -> sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/sdcflows/workflows/tests
Writing sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401/setup.cfg
Creating tar archive
removing 'sdcflows-1.2.1.dev5+gcdcdf2d1.d20200401' (and everything under it)
```
there must be some config in the setup.py flagging every file as data.

**Comment by oesteban** [comment]
I finally managed to get all tests to pass (and integrated CircleCI to show up under the Checks tab, although that is not working for previously existing PRs). I'm going to go ahead and cut a new release to see if everything works okay. We will also be able to check the github archives.

File: MANIFEST.in (hunk 2 above)
Comment (effigies) [reply MANIFEST.in]: I wonder if it's `setuptools_scm` deciding that everything under version control should go in the sdist.

**Comment by effigies** [comment]
It is the case that sdists package everything under version control by default. See https://github.com/pypa/setuptools_scm/issues/175, https://github.com/pypa/setuptools_scm/issues/199.

I'm not sure that this is actually a bad thing such that we need to exclude these files, unless we have particularly large files to exclude. We should also be uploading wheels, which is the preferred installation method when available.

**Pull request merged**